src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorViewCompiler.cs

```diff
@@ -34,16 +34,17 @@ public class RazorViewCompiler : IViewCompiler
         private readonly IFileProvider _fileProvider;
         private readonly RazorProjectEngine _projectEngine;
         private readonly Action<RoslynCompilationContext> _compilationCallback;
+        private readonly IMemoryCache _cache;
```

Comment:
I don't think it should be here, because you can't assume when the instance should be disposed. The provider is responsible for providing the instance, and is also the only one to be able to decide when it should be disposed. Disposing the compiler should not dispose the cache. In a multi-tenant case for instance you want to keep the cache alive even if one of the compilers is disposed.